### DISCUSSION TITLE
Error 0x6A80 for Applet Install on sysmoISIM-SJA2

### DISCUSSION THREAD

**Discussion by micahflack** [martinpaljak/GlobalPlatformPro#358] [Q&A]
Hello! Does anyone here know how to install applets to the sysmoISIM-SJA2 SIM cards? I wasn't sure if this was just a user error or a bug, so I started this discussion instead.

Normally, I would use this updated version of shadysim to install the applets: [herlesupreeth / sim-tools](https://github.com/herlesupreeth/sim-tools/tree/master)

But I am trying to make my applet the default and selectable, which GlobalPlatformPro should be able to do using this cmd?

`gp --key-dek ... --key-enc ... --key-mac ... --install ../HelloSTK2/bin/HelloSTK2.cap --default`

This is the output:

```
# GlobalPlatformPro v21.12.31-55-g52c3581
# Running on Linux 5.15.0-117-generic amd64, Java 11.0.23 by Ubuntu
[INFO] GPSession - Using card master keys with version 0 for setting up session with MAC 
[INFO] GPSession - Diversified card keys: ENC=ACCFA969C733F0EFC22CCB6E19D98CA1 (KCV: 3AB395) MAC=763D9708C2E42898FCE918D4928314F4 (KCV: 1F5A19) DEK=61F5DEEE493784FA3FC324C3C13309BA (KCV: 9E1AD2) for SCP02
[INFO] GPSession - Session keys: ENC=94B60BB2DA2420D0995D3129044F8E74 MAC=6E3C2D0D39090BDB51598804251B4DAF RMAC=423A37EF946710A35CD404EF751538D4
CAP file (v2.1), contains: applets for JavaCard 2.2.1
Package: com.github.mrlnc.HelloSTK2 D07002CA44 v1.0
Applet:  com.github.mrlnc.HelloSTK2.HelloSTK2 D07002CA44900101
Import:  A0000000620101                   v1.2 javacard.framework
Import:  A0000000090003FFFFFFFF8910710002 v2.6 sim.toolkit
Import:  A0000000620001                   v1.0 java.lang
Generated by Oracle Corporation converter  [v3.0.5]
On Fri Jul 26 03:12:12 MDT 2024 with JDK 11.0.23 (Ubuntu)
Code size 1104 bytes (1456 with debug)
SHA-256 93c33c45265c9eae9038a9e338be98fc8603b38aca4e3922ba0444dcedb9d541
SHA-1   eec489f79c6297d6f8617a6f9e7bd00c700b953c
../HelloSTK2/bin/HelloSTK2.cap loaded: com.github.mrlnc.HelloSTK2 D07002CA44
Error: INSTALL [for install and make selectable] failed: 0x6A80 (Wrong data/incorrect values in data)
```

My understanding is that this might be caused by GlobalPlatformPro using the wrong installation parameters? And, I need to specify those used by the card manufacturer? Maybe this is similar to a previous issue... [#293](https://github.com/martinpaljak/GlobalPlatformPro/issues/293)

I think I am headed in the right direction - but, I was hoping to see if someone else has knowledge/experience first.

**Comment by martinpaljak** [comment]
It is the right direction - SIM cards require specific parameters. The comments at linked ticket describes the path towards a solution as well.

**Comment by mrlnc** [comment]
I was trying to use ShadySIM to generate the Application Parameters TLV and then LOAD and INSTALL via GPPro. This failed. Cross-posting [my notes from the Osmocom tracker](https://osmocom.org/issues/6069#note-12):

GPPro assumes that within the params TLV the tag 0xc9 (Application Specific Parameters) always comes first (see [GPSession.java](https://github.com/martinpaljak/GlobalPlatformPro/blob/52c3581cd8309f4dac230d21886e1bf69d362138/library/src/main/java/pro/javacard/gp/GPSession.java#L653C39-L653C39)). Shadysim however puts the element 0xEF first, then 0xC9. ([link to code](https://github.com/herlesupreeth/sim-tools/blob/c7110a85344f9688591cf5a5853bfc962f9a63eb/shadysim/shadysim_isim.py#L393C29-L393C29)). In that case, GPPro prepends another 0xC9 tag, invalidating the whole TLV.

So just swap the elements in [this line](https://github.com/herlesupreeth/sim-tools/blob/c7110a85344f9688591cf5a5853bfc962f9a63eb/shadysim/shadysim_isim.py#L393C29-L393C29) and print parameters:
```python
- parameters = 'ef' + ('%02x' % (len(parameters) / 2)) + parameters + 'c9' + ('%02x' % (len(args.app_parameters) / 2)) + args.app_parameters
+ parameters = 'c9' + ('%02x' % (len(args.app_parameters) / 2)) + args.app_parameters + 'ef' + ('%02x' % (len(parameters) / 2)) + parameters
```

Params for ImsiChange that seemed to work:
```
$ java -jar ./gp.jar --key-enc XXX --key-mac XXX --key-dek XXX --install  ../hello-stk/build/ImsiChange.cap --params c900ef2ac80201FFc70201FFca20010001000f0c0000000000000000000000000000000000000000000000000000
CAP loaded
```

**Reply by martinpaljak** [reply]
This sounds like GPPro being a bit naive and this could be considered a bug. https://github.com/martinpaljak/GlobalPlatformPro/issues/360

**Comment by martinpaljak** [comment]
fyi, this should be fixed in master branch
